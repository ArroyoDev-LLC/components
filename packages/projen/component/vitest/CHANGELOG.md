# Changelog

## 1.0.0 (2023-05-01)


### Features

* **projen.component.vitest:** Migrate vitest to component package. ([ce08c4d](https://github.com/ArroyoDev-LLC/components/commit/ce08c4dbd5a306dab739da56c9f0f0529c78ec49))
* **projen:** Update managed files. ([964ade5](https://github.com/ArroyoDev-LLC/components/commit/964ade56809db26a69c569eab4d3520cdb30c93f))


### Bug Fixes

* **utils.projen:** Unused type ([9141769](https://github.com/ArroyoDev-LLC/components/commit/91417699cfbdf16c57fcfb2caea522bc4ece785d))
