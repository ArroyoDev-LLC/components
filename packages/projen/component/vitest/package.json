{
  "name": "@arroyodev-llc/projen.component.vitest",
  "scripts": {
    "build": "pnpm exec projen build",
    "bump": "pnpm exec projen bump",
    "clean": "pnpm exec projen clean",
    "compile": "pnpm exec projen compile",
    "default": "pnpm exec projen default",
    "eslint": "pnpm exec projen eslint",
    "package": "pnpm exec projen package",
    "post-compile": "pnpm exec projen post-compile",
    "post-install": "pnpm exec projen post-install",
    "post-upgrade": "pnpm exec projen post-upgrade",
    "pre-compile": "pnpm exec projen pre-compile",
    "release": "pnpm exec projen release",
    "stub": "pnpm exec projen stub",
    "test": "pnpm exec projen test",
    "unbump": "pnpm exec projen unbump",
    "upgrade": "pnpm exec projen upgrade",
    "watch": "pnpm exec projen watch"
  },
  "author": {
    "name": "arroyoDev-LLC",
    "email": "support@arroyodev.com",
    "url": "https://arroyodev.com",
    "organization": true
  },
  "devDependencies": {
    "@arroyodev-llc/utils.unbuild-composite-preset": "^0.1.3",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "eslint": "^8",
    "eslint-config-prettier": "^9.0.0",
    "eslint-import-resolver-node": "^0.3.9",
    "eslint-import-resolver-typescript": "^3.6.1",
    "eslint-plugin-import": "^2.28.1",
    "eslint-plugin-prettier": "^5.0.1",
    "lint-staged": "^14.0.1",
    "npm-check-updates": "^16",
    "prettier": "^3.0.3",
    "projen": "0.76.0",
    "standard-version": "^9",
    "typescript": "~5.1",
    "unbuild": "^1.2.1"
  },
  "peerDependencies": {
    "projen": "^0.76.0"
  },
  "dependencies": {
    "@arroyodev-llc/projen.component.typescript-source-file": "workspace:*",
    "@arroyodev-llc/projen.component.vite": "workspace:*",
    "@arroyodev-llc/utils.projen": "workspace:*",
    "@arroyodev-llc/utils.ts-ast": "workspace:*",
    "ts-morph": "^19.0.0",
    "vitest": "^0.34.6"
  },
  "main": "dist/index.mjs",
  "license": "Apache-2.0",
  "publishConfig": {
    "access": "public"
  },
  "version": "0.1.23",
  "types": "./dist/index.d.ts",
  "type": "module",
  "sideEffects": false,
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    },
    "./package.json": "./package.json"
  },
  "files": [
    "dist"
  ],
  "packageManager": "pnpm@8.9.2",
  "lint-staged": {
    "*.{ts,tsx,mts,cts,ts,tsx,mts,cts}": [
      "pnpm exec eslint --no-error-on-unmatched-pattern --fix"
    ],
    "*.{yaml,yml}": [
      "pnpm exec prettier --write"
    ]
  },
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
