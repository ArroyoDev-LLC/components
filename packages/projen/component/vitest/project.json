{
  "name": "projen.component.vitest",
  "root": "packages/projen/component/vitest",
  "targets": {
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "build": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "bump": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen bump",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "unbump": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen unbump",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "release": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen release",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "post-upgrade": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-upgrade",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "upgrade": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen upgrade",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "stub": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen stub",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/projen/component/vitest"
      }
    },
    "clean": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen clean",
        "cwd": "packages/projen/component/vitest"
      }
    }
  },
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
