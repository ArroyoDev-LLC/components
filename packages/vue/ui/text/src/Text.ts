import { computed, defineComponent, h } from 'vue';

type HeaderNumbers = 1 | 2 | 3 | 4 | 5 | 6;
type HeaderNames = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export type TextVariant = HeaderNumbers | HeaderNames | 'body';

export interface TextProps {
  variant?: TextVariant;
}

export default defineComponent<TextProps>({
  name: 'BaseText',
  setup(props, { slots }) {
    const header = computed(() => props.variant ?? 'body');
    const anchor = computed(() =>
      header.value === 'body'
        ? 'p'
        : typeof header.value === 'number'
          ? `h${header.value}`
          : header.value,
    );
    return () => h(anchor.value, {}, slots.default?.());
  },
});
