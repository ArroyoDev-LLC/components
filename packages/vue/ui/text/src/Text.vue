<script setup lang="ts">
import { computed } from 'vue';
import type { TextVariant } from './types.ts';

export interface TextProps {
  variant?: TextVariant;
}

const props = withDefaults(defineProps<TextProps>(), { variant: 'body' });

const anchor = computed(() =>
  props.variant === 'body'
    ? 'p'
    : typeof props.variant === 'number'
      ? `h${props.variant}`
      : props.variant,
);
</script>

<template>
  <component :is="anchor">
    <slot />
  </component>
</template>
