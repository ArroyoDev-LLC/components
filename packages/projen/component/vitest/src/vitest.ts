import path from 'node:path'
import { TypeScriptSourceConfig } from '@arroyodev-llc/projen.component.typescript-source-file'
import { Vite } from '@arroyodev-llc/projen.component.vite'
import { cwdRelativePath } from '@arroyodev-llc/utils.projen'
import { Component, JsonFile, type Project, type Task } from 'projen'
import { type TypeScriptProject } from 'projen/lib/typescript'
import { deepMerge } from 'projen/lib/util'
import { type UserWorkspaceConfig } from 'vitest/config'

export enum VitestConfigType {
	WORKSPACE = 'Workspace',
	PROJECT = 'Project',
	CONFIG = 'Config',
}

export interface VitestOptions {
	/**
	 * Vitest config path.
	 */
	configFilePath?: string
	/**
	 * Configuration type.
	 */
	configType?: VitestConfigType
	/**
	 * Vitest settings.
	 */
	settings?: UserWorkspaceConfig
}

export class Vitest extends Component {
	public static of(project: Project): Vitest | undefined {
		const isVitest = (c: Component): c is Vitest => c instanceof Vitest
		return project.components.find(isVitest)
	}

	readonly configType: VitestConfigType
	readonly configFile: TypeScriptSourceConfig<UserWorkspaceConfig>
	readonly options: Required<VitestOptions>
	readonly vite?: Vite
	readonly testWatchTask: Task
	readonly #workspaceProjects: Map<string, Vitest> = new Map<string, Vitest>()

	constructor(
		public readonly project: TypeScriptProject,
		options: VitestOptions = {
			configFilePath: 'vitest.config.ts',
			configType: VitestConfigType.PROJECT,
			settings: {},
		}
	) {
		super(project)
		this.vite = Vite.of(this.project)
		this.options = deepMerge([
			{
				settings: {
					test: {
						name: project.name,
						include: [`${project.testdir}/\*\*/\*.spec.ts`],
					},
				},
			} as VitestOptions,
			options,
		]) as Required<VitestOptions>
		this.project.addDevDeps('vitest')
		this.configType = this.options.configType ?? VitestConfigType.PROJECT
		const configFilePath = this.options.configFilePath ?? 'vitest.config.ts'

		if (this.configType !== VitestConfigType.WORKSPACE) {
			this.project.testTask.exec('vitest', {
				args: ['--run'],
				receiveArgs: true,
			})
		}
		this.testWatchTask =
			this.project.tasks.tryFind('test:watch') ??
			this.project.addTask('test:watch', {
				receiveArgs: true,
				exec: 'vitest',
				description: 'Run tests on changes.',
			})

		this.project.tsconfigDev?.addInclude?.(configFilePath)
		this.project?.eslint?.addOverride?.({
			files: [
				configFilePath,
				...(this.options.settings?.test?.include ?? [this.project.testdir]),
			],
			rules: {
				'import/no-extraneous-dependencies': ['off'],
			},
		})

		this.configFile =
			TypeScriptSourceConfig.withCallExpressionConfig<UserWorkspaceConfig>(
				this.project,
				configFilePath,
				{
					source: `export default ${this.defineName}({})`,
					recreate: true,
					marker: true,
					config: this.options.settings,
				}
			)

		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: [this.defineName],
		})

		if (this.configType === VitestConfigType.PROJECT) {
			this.tryFindWorkspace()?.addProjectConfig?.(this)
		}

		if (this.configType === VitestConfigType.WORKSPACE) {
			new JsonFile(this.project, 'vitest.workspace.json', {
				readonly: true,
				allowComments: false,
				obj: () => Array.from(this.#workspaceProjects.keys()),
			})
		}
	}

	/**
	 * Get the define function name.
	 * @protected
	 */
	protected get defineName(): string {
		return this.configType === VitestConfigType.PROJECT
			? 'defineProject'
			: 'defineConfig'
	}

	/**
	 * Merge vite config if applicable.
	 * @protected
	 */
	protected mergeVite(): void {
		if (!this.vite) return
		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: ['mergeConfig'],
		})
		this.configFile.addImport({
			moduleSpecifier: cwdRelativePath(
				this.project.outdir,
				this.vite.file.absolutePath.split('.').slice(0, -1).join('.')
			),
			defaultImport: 'viteConfig',
		})
		this.configFile.addConfigTransform((cfg, src) => {
			src.addExportAssignment({
				expression: (writer) =>
					writer.write(
						`mergeConfig(viteConfig, ${this.defineName}(${cfg.getFullText()}))`
					),
				isExportEquals: false,
			})
			src.removeDefaultExport()
		})
	}

	/**
	 * Try to find the workspace {@link Vitest} component.
	 */
	tryFindWorkspace(): Vitest | undefined {
		if (this.configType === VitestConfigType.WORKSPACE) return this
		if (this.project.parent) {
			return Vitest.of(this.project.parent)?.tryFindWorkspace?.()
		}
		return
	}

	/**
	 * Add {@link Vitest} component to this workspace.
	 * @param component Target component.
	 */
	addProjectConfig(component: Vitest): this {
		if (this.configType !== VitestConfigType.WORKSPACE)
			throw new Error('No vitest workspace found. Cannot add project config!')
		const relPath = path.format({
			dir: '.',
			base: path.relative(
				path.dirname(this.configFile.absolutePath),
				component.configFile.absolutePath
			),
		})
		this.#workspaceProjects.set(relPath, component)
		return this
	}

	/**
	 * Add config to merge.
	 * @param config Config part.
	 */
	addConfig(config: this['configFile']['__mergeSchema']): this {
		this.configFile.addConfig(config)
		return this
	}

	/**
	 * Add plugin to merge.
	 * @param plugin Plugin part.
	 */
	addPlugin(plugin: this['configFile']['__pluginSchema']) {
		this.configFile.addPlugin(plugin)
		return this
	}

	/**
	 * @inheritDoc
	 */
	preSynthesize() {
		super.preSynthesize()
		this.mergeVite()
	}
}
