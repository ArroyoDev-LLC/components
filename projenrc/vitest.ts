import path from 'node:path'
import { Component, JsonFile, type Project } from 'projen'
import { type TypeScriptProject } from 'projen/lib/typescript'
import {
	type ObjectLiteralExpression,
	type SourceFile,
	SyntaxKind,
} from 'ts-morph'
import { type UserWorkspaceConfig } from 'vitest/config'
import {
	TypeScriptSourceFile,
	type TypeScriptSourceFileTransform,
} from '@arroyodev-llc/projen.component.typescript-source-file'
import { addPropertyAssignmentsFromObject } from '@arroyodev-llc/utils.projen'

export enum VitestConfigType {
	WORKSPACE = 'Workspace',
	PROJECT = 'Project',
	CONFIG = 'Config',
}

export interface VitestOptions {
	configFilePath?: string
	configType?: VitestConfigType
	settings?: UserWorkspaceConfig
}

export class Vitest extends Component {
	public static of(project: Project): Vitest | undefined {
		const isVitest = (c: Component): c is Vitest => c instanceof Vitest
		return project.components.find(isVitest)
	}

	readonly configType: VitestConfigType
	readonly configFile: TypeScriptSourceFile
	readonly #workspaceProjects: Map<string, Vitest> = new Map<string, Vitest>()

	constructor(
		public readonly project: TypeScriptProject,
		public readonly options: VitestOptions = {
			configFilePath: 'vitest.config.ts',
			configType: VitestConfigType.PROJECT,
			settings: {},
		}
	) {
		super(project)
		this.configType = this.options.configType ?? VitestConfigType.PROJECT
		const configFilePath = this.options.configFilePath ?? 'vitest.config.ts'

		this.project.tsconfigDev?.addInclude?.(configFilePath)
		this.project?.eslint?.addOverride?.({
			files: [
				configFilePath,
				...(this.options.settings?.test?.include ?? [this.project.testdir]),
			],
			rules: {
				'import/no-extraneous-dependencies': ['off'],
			},
		})

		this.configFile = new TypeScriptSourceFile(this.project, configFilePath, {
			source: `export default ${this.defineName}({})`,
			recreate: true,
		})

		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: [this.configType],
		})

		if (this.configType === VitestConfigType.PROJECT) {
			this.tryFindWorkspace()?.addProjectConfig?.(this)
		}

		if (this.configType === VitestConfigType.WORKSPACE) {
			new JsonFile(this.project, 'vitest.workspace.json', {
				readonly: true,
				allowComments: false,
				obj: () => Array.from(this.#workspaceProjects.keys()),
			})
		}

		this.addConfigTransform((configExpr) => {
			const { test = { include: ['test/**.spec.ts'] }, ...rest } =
				this.options.settings ?? {}
			addPropertyAssignmentsFromObject(configExpr, { test, ...rest })
		})
	}

	protected get defineName(): string {
		return this.configType === VitestConfigType.PROJECT
			? 'defineProject'
			: 'defineConfig'
	}

	tryFindWorkspace(): Vitest | undefined {
		if (this.configType === VitestConfigType.WORKSPACE) return this
		if (this.project.parent) {
			return Vitest.of(this.project.parent)?.tryFindWorkspace?.()
		}
		return
	}

	addProjectConfig(component: Vitest): this {
		if (this.configType !== VitestConfigType.WORKSPACE)
			throw new Error('No vitest workspace found. Cannot add project config!')
		const relPath = path.format({
			dir: '.',
			base: path.relative(
				path.dirname(this.configFile.absolutePath),
				component.configFile.absolutePath
			),
		})
		this.#workspaceProjects.set(relPath, component)
		return this
	}

	addConfigTransform(
		transform: (
			configObjectLiteral: ObjectLiteralExpression,
			sourceFile: SourceFile
		) => void
	) {
		const transformer: TypeScriptSourceFileTransform = (src: SourceFile) => {
			const configExport = src
				.getExportAssignmentOrThrow((exp) =>
					Boolean(exp.getExpressionIfKind(SyntaxKind.CallExpression))
				)
				.getExpressionIfKindOrThrow(SyntaxKind.CallExpression)
			const configExpr = configExport
				.getArguments()[0]!
				.asKindOrThrow(SyntaxKind.ObjectLiteralExpression)
			transform(configExpr, src)
		}
		this.configFile.addTransformer(transformer)
		return this
	}
}
