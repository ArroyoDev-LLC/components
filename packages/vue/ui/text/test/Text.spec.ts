import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import Text from '../src/Text.vue'

describe('Text', () => {
	it('renders as expected.', async () => {
		const mnt = mount(Text)
		expect(mnt.element).toMatchSnapshot()
	})

	it.each([1, 2, 3, 4, 5, 6, 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'])(
		'renders all variants as expected.',
		async (variant) => {
			const mnt = mount(Text, { props: { variant }, slots: { default: 'Hi' } })
			expect(mnt.element).toMatchSnapshot()
		}
	)
})
