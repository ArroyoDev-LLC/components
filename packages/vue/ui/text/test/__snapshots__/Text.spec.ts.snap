// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Text > renders all variants as expected. 1`] = `
<h1>
  
  Hi
  
</h1>
`;

exports[`Text > renders all variants as expected. 2`] = `
<h2>
  
  Hi
  
</h2>
`;

exports[`Text > renders all variants as expected. 3`] = `
<h3>
  
  Hi
  
</h3>
`;

exports[`Text > renders all variants as expected. 4`] = `
<h4>
  
  Hi
  
</h4>
`;

exports[`Text > renders all variants as expected. 5`] = `
<h5>
  
  Hi
  
</h5>
`;

exports[`Text > renders all variants as expected. 6`] = `
<h6>
  
  Hi
  
</h6>
`;

exports[`Text > renders all variants as expected. 7`] = `
<h1>
  
  Hi
  
</h1>
`;

exports[`Text > renders all variants as expected. 8`] = `
<h2>
  
  Hi
  
</h2>
`;

exports[`Text > renders all variants as expected. 9`] = `
<h3>
  
  Hi
  
</h3>
`;

exports[`Text > renders all variants as expected. 10`] = `
<h4>
  
  Hi
  
</h4>
`;

exports[`Text > renders all variants as expected. 11`] = `
<h5>
  
  Hi
  
</h5>
`;

exports[`Text > renders all variants as expected. 12`] = `
<h6>
  
  Hi
  
</h6>
`;

exports[`Text > renders as expected. 1`] = `
<p>
  
  
</p>
`;
